### PR TITLE
Add helper functions for optlang solvers that work with new context

### PR DESCRIPTION
**Pull request by cdiener**
This is a proposal to pave the way for the optlang port of `flux_analysis.py`. The functions defined in `cobra.util.solver` help to temporarily add or remove variables and constraints to the solver. It also includes helpers to extract all available solver interfaces from `optlang`.

This implied some minor changes in `cobra.Model` in setting `Model.solver` and `Model.optimize`. 

The "solver" keyword is now respected in `Model.optimize` but requires a "optlang-" prefix for the optlang solvers.

### CODE REVIEW

**Comment by cdiener** [comment]
The codecov thing is incorrect, see comment on gitter.

File: cobra/util/solver.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    """
+    context = get_context(model)
+
+    if len(what) > 0:
+        model.solver.add(what)
+        if context:
+            context(partial(model.solver.remove, what))
+
+
+def remove_from_solver(model, what=[]):
```
Comment (pstjohn) [inline cobra/util/solver.py]: Would we ever want to *not* check for a context? I.e., should we just be replacing all calls to model.solver.add() and model.solver.remove() with a context'ed version?

File: cobra/util/solver.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    # variable > -expression
+    constraints = [
+        # positive value constraint
+        model.solver.interface.Constraint(expression - variable, ub=0,
+                                          name="abs_pos_" + name),
+        # negative value constraint
+        model.solver.interface.Constraint(expression + variable, lb=0,
+                                          name="abs_neg_" + name)
+        ]
+    model.solver.add(constraints + [variable])
```
Comment (pstjohn) [inline cobra/util/solver.py]: couldn't you just call `add_to_solver` above, and not duplicate the context code?

File: cobra/util/solver.py (hunk 1 above)
Comment (cdiener) [inline cobra/util/solver.py]: model.solver is an `optlang.Model` and `add` is defined and implemented in optlang :/

File: cobra/util/solver.py (hunk 2 above)
Comment (cdiener) [inline cobra/util/solver.py]: You're right! Silly me :) will change it

File: cobra/util/solver.py (hunk 1 above)
Comment (pstjohn) [inline cobra/util/solver.py]: right, but we could replace all calls to `model.solver.add()` with `add_to_solver`. We could even consider making it a cobra.Model method, `model.add_to_solver(self, ...)`

Comment (cdiener) [inline cobra/util/solver.py]: Ah ok, I misunderstood. Yes, that would be feasible. Also thought about making it methods of `Model` but remembered that you were against adding too many new functions in `Model.py` :smile:

**Pull request merged**